prov/sockets/src/sock_poll.c

```diff
@@ -141,13 +141,19 @@ static int sock_poll_poll(struct fid_poll *pollset, void **context, int count)
 			cntr = container_of(list_item->fid, struct sock_cntr,
 						cntr_fid);
 			sock_cntr_progress(cntr);
-			pthread_mutex_lock(&cntr->mut);
-			if (atomic_get(&cntr->value) >=
-				atomic_get(&cntr->threshold)) {
-				*context++ = cntr->cntr_fid.fid.context;
-				ret_count++;
+			{
+				int val = atomic_get(&cntr->value);
+				int old_val = atomic_get(&cntr->old_value);
+				if (val > old_val) {
+					*context++ = cntr->cntr_fid.fid.context;
+					ret_count++;
+					int new_val = old_val;
+					do {
+						old_val = new_val;
+						new_val = atomic_cswap(&cntr->old_value, old_val, val);
+					} while(new_val != old_val && new_val < val);
+				}
```

Comment:
The change this patch makes to the sock_cntr structure (removing threshold) invalidates the existing fi_poll implementation. We could leave threshold in the struct and have a dead code path (since no threshold would ever be set), or return -FI_EINVAL in this branch.

This part of the patch implements a reasonable (I think) interpretation of fi_poll's semantics by returning any counters whose values have changed since the last fi_poll call -- consistent with the CQ and EQ branches returning whenever there is new data available. If it is better to leave dead code or return -FI_EINVAL, I can update the patch accordingly.